### PR TITLE
[Feat/backend refactoring] backend.py general refactoring

### PR DESCRIPTION
**Pull request by PascalDR**
This PR is relative to [212](https://github.com/italia/eudi-wallet-it-python/issues/212) and apply a complete refactoring to the satosa package.

### CODE REVIEW

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
+import uuid
+import json
+from typing import Callable
+from pyeudiw.jwk import JWK
+from satosa.context import Context
+from pydantic import ValidationError
+from ..impl.trust import BackendTrust
```
Comment (peppelinux) [inline pyeudiw/satosa/default/openid4vp_backend.py:7]: please do not use relative imports, like .. or . 

we use absolute imports only

Comment (peppelinux) [reply pyeudiw/satosa/default/openid4vp_backend.py:7]: please change the name impl to something more eloquent

what's the purpose of the word `impl`?

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+from ..impl.response import JsonResponse
+from pyeudiw.tools.utils import iat_now
+from satosa.internal import InternalData
+from pyeudiw.storage.db_engine import DBEngine
+from satosa.response import Response, Redirect
+from pyeudiw.tools.mobile import is_smartphone
+from urllib.parse import quote_plus, urlencode
+from ..impl.html_template import Jinja2TemplateHandler
+from pyeudiw.storage.exceptions import StorageWriteError
+from pyeudiw.federation.schemas.wallet_relying_party import WalletRelyingParty
```
Comment (peppelinux) [inline pyeudiw/satosa/default/openid4vp_backend.py:17]: please use `isort` to sort the imports by name

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        :rtype: satosa.response.Response
+        """
+
+        self._log_function_debug(
+            "pre_request_endpoint", context, "internal_request", internal_request)
+
+        session_id = context.state["SESSION_ID"]
+        state = str(uuid.uuid4())
+
+        # TODO: do not init the session if the context is not linked to any
```
Comment (peppelinux) [inline pyeudiw/satosa/default/openid4vp_backend.py:145]: please update the related issue when this PR will be approved, to match the right line number and facilitate the tracking

**Review by peppelinux — CHANGES_REQUESTED** [review]
@PascalDR good direction, several changes requested, let's continue in this path

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        for k, v in self.config['metadata'].items():
+            if isinstance(v, (int, float, dict, list)):
+                continue
+            if not v or len(v) == 0:
+                continue
+            if all((
+                v[0] == '<',
+                v[-1] == '>',
+                '.' in v
+            )):
```
[the remark this answers is not in the document]
Comment (peppelinux) [reply pyeudiw/satosa/default/openid4vp_backend.py:291]: here an example @PascalDR about how to use refs in yaml

https://github.com/UniversitaDellaCalabria/SATOSA-oidcop/pull/49/files#diff-f8c795b9eed9e666eb777dc5679d8886f2245826ea73186730408dfaeee54611R5

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/satosa/default/request_handler.py [hunk 5]
```diff
@@ -25,6 +25,35 @@
...
+
+            # for each single vp token, take the credential within it, use cnf.jwk to validate the vp token signature -> if not exception
+            # establish the trust to each credential issuer
+            tchelper = self._validate_trust(context, vp.payload['vp'])
+
+            if not tchelper.is_trusted:
+                return self._handle_400(context, f"Trust Evaluation failed for {tchelper.entity_id}")
+
+            # TODO: generalyze also for x509
+            credential_jwks = tchelper.get_trusted_jwks(
```
Comment (peppelinux) [inline pyeudiw/satosa/default/request_handler.py:42]: trust and credentials are different things, then this method is related to credential evaluation and trust evaluation to their issuers

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 6]
```diff
@@ -200,7 +200,11 @@ def get_response_endpoint(self, context: Context) -> Response:
         self._log_function_debug("get_response_endpoint", context)
 
         state = context.qs_params.get("id", None)
-        session_id = context.state["SESSION_ID"]
+        session_id = context.state.get("SESSION_ID", None)
+
+        if not state:
+            return self._handle_400(context, "No state found in query params")
```
Comment (peppelinux) [inline pyeudiw/satosa/default/openid4vp_backend.py:212]: SESSION_ID is not provided in the query params but in the internal storage instance bound to the cookie brought by the user-agent

please say that the SESSION ID is not found or something like this

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/satosa/default/request_handler.py (hunk 5 above)
Comment (PascalDR) [reply pyeudiw/satosa/default/request_handler.py:42]: Could "_handle_credential_trust" be a good name?

File: pyeudiw/satosa/default/request_handler.py [hunk 7]
```diff
@@ -142,7 +142,7 @@ def request_endpoint(self, context: Context, *args: tuple) -> Redirect | JsonRes
             # TODO - what if the credential is different from sd-jwt? -> generalyze within Vp class
 
             try:
-                vp.verify_sdjwt(
+                vp.verify(
                     issuer_jwks_by_kid={
```
Comment (peppelinux) [inline pyeudiw/satosa/default/request_handler.py:146]: mdoc cbor non usa i jwks, attento. Questo non avrebbe senso

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: pyeudiw/openid4vp/direct_post_response.py [hunk 8]
```diff
@@ -146,7 +147,7 @@ def get_presentation_vps(self) -> list[Vp]:
 
         for vp in vps:
             # TODO - add an exception handling here
-            _vp = Vp(vp)
+            _vp = handle_VP(vp)
```
Comment (peppelinux) [inline pyeudiw/openid4vp/direct_post_response.py:150]: non è un nome di compliant alla PEP, rinominalo in vp_verifier, vp_parser o simile

File: pyeudiw/openid4vp/utils.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+from pyeudiw.openid4vp.vp import Vp
+from pyeudiw.openid4vp.vp_sd_jwt import VpSdJwt
+from pyeudiw.jwt.utils import decode_jwt_header
+from pyeudiw.openid4vp.exceptions import VPFormatNotSupported
+
+
+def handle_VP(jwt: str) -> Vp:
```
Comment (peppelinux) [inline pyeudiw/openid4vp/utils.py:7]: non è un nome di compliant alla PEP, rinominalo in vp_verifier, vp_parser o simile

File: pyeudiw/satosa/default/openid4vp_backend.py (hunk 3 above)
Comment (peppelinux) [reply pyeudiw/satosa/default/openid4vp_backend.py:159]: pelase add a logger.warning by saying that this feature is not implemented yet

File: pyeudiw/satosa/default/request_handler.py (hunk 5 above)
Comment (peppelinux) [reply pyeudiw/satosa/default/request_handler.py:42]: yeah, we can do that

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+      class: DefaultOpenID4VPBackend
+    response_class: 
+      module: pyeudiw.satosa.default.response_handler
+      class: DefaultResponseHandler
+    request_class: 
+      module: pyeudiw.satosa.default.request_handler
+      class: DefaultRequestHandler
+  
+  ui:
+    static_storage_url: "https://localhost"
```
Comment (peppelinux) [inline example/satosa/pyeudiw_dynamic_backend.yaml:19]:
```suggestion
    static_storage_url: "https://localhost/static"
```

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 11]
```diff
@@ -0,0 +1,221 @@
+module: pyeudiw.satosa.dynamic_backend.DynamicBackend
+name: OpenID4VP
+
+config:
+
+  dynamic_backend:
```
Comment (peppelinux) [inline example/satosa/pyeudiw_dynamic_backend.yaml:6]: dyanmic_backed doesn't means nothing. According to what we have said during our alignment meeting, each class MUST be configured within the corresponding endpoint, in the section where the endpoints are defined

````
  endpoints:
      pre_request: 
        class: pyeudiw.satosa.default.request_handler
        module: OpenID4VPBackend
        path: '/pre-request'
      ...
````

and so on. 
Please remove "Default" from all the names.

Comment (peppelinux) [reply example/satosa/pyeudiw_dynamic_backend.yaml:6]: the base class is the one of the entire backend
https://github.com/italia/eudi-wallet-it-python/pull/229/commits/1e3c9fb685bad6f9fbb51680f40f132a21b21e40#diff-cd382c85588277c666bbd8c9e5c32a931650eab9e0f44188a52a5c2eae28b2f2R1

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        - RS384
+        - RS512
+        - ES256
+        - ES384
+        - ES512
+    
+  authorization:
+    url_scheme: eudiw
+    scopes: 
+    - pid-sd-jwt:unique_id+given_name+family_name
```
Comment (peppelinux) [inline example/satosa/pyeudiw_dynamic_backend.yaml:70]: please align this value with the current eudi-wallet-id-docs

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    federation_entity_metadata:
+        organization_name: Developers Italia SATOSA OpenID4VP backend
+        homepage_uri: https://developers.italia.it
+        policy_uri: https://developers.italia.it
+        tos_uri: https://developers.italia.it
+        logo_uri: https://developers.italia.it/assets/icons/logo-it.svg
+
+    # private jwk
+    federation_jwks:
+      - kty: RSA
```
Comment (peppelinux) [inline example/satosa/pyeudiw_dynamic_backend.yaml:103]: please switch to EC, we abandon RSA for now

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+    # loaded in the __init__
+    # jwks:
+
+    #This section contains the details for presentation request
+    presentation_definition:
+      id: d76c51b7-ea90-49bb-8368-6b3d194fc131
+      input_descriptors:
+        - id: IdentityCredential
+          format:
+            vc+sd-jwt: { }
```
Comment (peppelinux) [inline example/satosa/pyeudiw_dynamic_backend.yaml:191]: why this is an empty {}?

**Review by peppelinux — CHANGES_REQUESTED** [review]

File: example/satosa/pyeudiw_dynamic_backend.yaml [hunk 15]
```diff
@@ -0,0 +1,221 @@
+module: pyeudiw.satosa.dynamic_backend.DynamicBackend
```
[the remark this answers is not in the document]
Comment (peppelinux) [reply example/satosa/pyeudiw_dynamic_backend.yaml:1]: the yaml filename of the backend MUST be the old one as well.

File: pyeudiw/satosa/default/openid4vp_backend.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+        :type config: dict[str, dict[str, str] | list[str]]
+        :param base_url: base url of the service
+        :type base_url: str
+        :param name: name of the plugin
+        :type name: str
+        """
+        super().__init__(auth_callback_func, internal_attributes, base_url, name)
+
+        self.config = config
+        self._client_id = self.config['metadata']['client_id']
```
[the remark this answers is not in the document]
Comment (PascalDR) [reply pyeudiw/satosa/default/openid4vp_backend.py:52]: It has the public getter on line [339](https://github.com/italia/eudi-wallet-it-python/pull/229/files/3c2aa8eac58636c4f9009a5b74c6e0930d9423db#diff-046cdbbd6ab4bcb52a0bef3a8ec01a53a2c530631d96ea9bbc65418d57e30368R339).
Do you prefer how was in the original?

**Review by peppelinux — APPROVED** [review]
It will be merged after the tag of the new release 0.8.0

**Pull request merged**